examples/express-graphql-http-functions/Query.ts

```diff
@@ -0,0 +1,73 @@
+import { Context } from "./context";
+import User from "./models/User";
+
+/**
+ * @gqlType
+ */
+export type Query = unknown;
+
+
+/**
+ * @gqlField users
+ */
+export function usersQueries(_: Query): UserQueries {
+  return {}
+}
+
+/**
+ * @gqlType
+ */
+export type UserQueries = {};
+
+/** @gqlField */
+export async function list(_: UserQueries, _args: {}, { userService }: Context): Promise<User[]> {
+  return userService.listUsers().then(users => users.map(user => new User(user.id, user)));
+}
+
+/** @gqlField */
+export async function byId(_: UserQueries, { id }: { id: string }, { userService }: Context): Promise<User> {
+  return userService.getUser(id).then(user => new User(user.id, user))
+}
+
+/**
+ * @gqlType
+ */
+export type Mutation = unknown;
+
+/**
+ * @gqlField users
+ */
+export function usersMutations(_: Mutation): UserMutations {
```

Comment:
I thought there was an expectation that the actual side-effect of the mutation would happen in the resolver of the root field, rather than child feeds. In other words, wouldn't this be an anti-pattern?

I believe GraphQL enforces that you only have one top level Mutation field to ensure each request maps to one side-effect and this would interfere with that.

Maybe we need to track down the spec to see if I'm hallucinating here.

Reply:
I was hallucinating about GraphQL enforcing only one top-level mutation field (that's true for subscriptions).

But it does seem that there's some expectation that only the top level mutation fields will perform side effects. The only language I was able to find in the spec to confirm this was that there's sepecial language asserting the unlike every other selection set, the GraphQL executor will evaluate the top level selection set of a mutation serially in an attempt to make side-effect ordering stable/predictable.

See:
* https://spec.graphql.org/draft/#sel-EANPHDDAhFDFA6Gk7I
* https://spec.graphql.org/draft/#sec-Normal-and-Serial-Execution

So, if you move your side effect resolvers deeper you will lose the guarantee that your side effects will always apply in the same order. I'm sure for many (most) use cases this will not be a problem, but I think we ought to avoid demonstrating it as an emulatable pattern in Grat's examples.

Interesting topic though! I learned something today!

Reply:
As I see it this "poor's man namespace" pattern for mutations doesn't contradict the serial nature of the mutations execution (it would go leaf first), and in practice, yeah, it has never been a problem and works as expected.

However, I agree with you, this is something that's executor-specific, outside of the scope of Grats (I don't think Grats can influence the executor that way, right?) So I'll "flatten" all of this.

IMO this is a "bug"/oversight in the spec itself, don't you think? There should be clear language about this topic:
* Allowing _only_ top level mutation fields to perform side-effects (boooo)
* Allow something like this (n-level fields can perform side-effects) and define how should the executor go about performing these side effects.

Reply:
This sparked some interesting conversations in Discord chat. We eventually found the explicit declaration of this in the spec:

> the resolution of fields other than top-level mutation fields must always be side effect-free and idempotent

https://spec.graphql.org/draft/#sel-GANRNDABiEBuHxyV